python/taichi/graph/_graph.py

```diff
@@ -50,23 +51,40 @@ def __init__(self, compiled_graph) -> None:
         self._compiled_graph = compiled_graph
 
     def run(self, args):
+        # Support native python numerical types (int, float), Ndarray.
+        # Taichi Matrix types are flattened into (int, float) arrays.
+        # TODO diminish the flatten behavior when Matrix becomes a Taichi native type.
         arg_ptrs = {}
-        # Only support native python numerical types (int, float) for now.
         arg_ints = {}
         arg_floats = {}
+        arg_doubles = {}
 
         for k, v in args.items():
             if isinstance(v, Ndarray):
                 arg_ptrs[k] = v.arr
             elif isinstance(v, int):
                 arg_ints[k] = v
             elif isinstance(v, float):
-                arg_floats[k] = v
+                arg_doubles[k] = v
+            elif isinstance(v, Matrix):
+                mat_val_id = 0
+                for a in range(v.n):
+                    for b in range(v.m):
+                        key = f"{k}_mat_{mat_val_id}"
+                        mat_val_id += 1
+                        if isinstance(v[a, b], int):
+                            arg_ints[key] = int(v[a, b])
+                        elif isinstance(v[a, b], float):
+                            arg_floats[key] = float(v[a, b])
```

Comment:
nit: Why is one going to `arg_doubles` and the other going to `arg_floats`..?

Reply:
The original implementation would cast to `double`. However my test case for Matrix is ti.f32 that should be cast to `float`, so use this branch for new matrix args and avoid break other args. The type needs further refactor, filed an issue on this #5165 .